File: Backend/pathfinder/demo/src/main/java/com/example/pathfinder/graph/Graph.java
package com.example.pathfinder.graph;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.PriorityQueue;

import com.example.pathfinder.utils.Utils;

public class Graph {
    private final Map<String, LinkedList<Edge>> adjacencyList;

    private final LinkedList<NodeCoordinate> graphCoordinates;

    public Graph() {

        this.adjacencyList = new HashMap<>();
        this.graphCoordinates = Utils.getCoordinatesFromCsv("node_coordinates.csv");
    }

    public void addEdge(Edge edge) {
        if (!adjacencyList.containsKey(edge.getSourceId())) {
            adjacencyList.put(edge.getSourceId(), new LinkedList<>());
        }

        if (!adjacencyList.containsKey(edge.getDestinationId())) {
            adjacencyList.put(edge.getDestinationId(), new LinkedList<>());
        }

        adjacencyList.get(edge.getSourceId()).add(edge);

        //Will also add the same path in reverse order for the graph to be bidirectional
        Edge reverseEdge = new Edge(edge.getDestinationId(), edge.getSourceId(), edge.getWeight());
        adjacencyList.get(edge.getDestinationId()).add(reverseEdge);
    }

    public LinkedList<String> getNodes() {
        return new LinkedList<>(adjacencyList.keySet());
    }

    public LinkedList<NodeCoordinate> getGraphCoordinates() {
        return graphCoordinates;
    }

    public Map<String, LinkedList<Map<String, Object>>> getGraph() {
        Map<String, LinkedList<Map<String, Object>>> graphContent = new HashMap<>();

        for (String key : adjacencyList.keySet()) {
            if (!graphContent.containsKey(key)) {
                graphContent.put(key, new LinkedList<>());
            }

            for (Edge edge : adjacencyList.get(key)) {
                graphContent.get(key).add(edge.toMap());
            }
        }

        return graphContent;
    }

    public String printGraph() {
        int i = 1;
        StringBuilder text = new StringBuilder();

        for (String k : adjacencyList.keySet()) {
            for (Edge edge : adjacencyList.get(k)) {
                text.append(String.format("vertex-%d %s\n", i, edge.toString()));
            }
            i++;
        }

        String text_string = text.toString();
        System.out.println(text_string);

        return text_string;
    }

    public Path dijkstra(String startVertex, String endVertex) {
        PriorityQueue<NodeDistance> priorityQueue = new PriorityQueue<>();
        Map<String, Boolean> visited = new HashMap<>();
        Map<String, Integer> distance = new HashMap<>();
        Map<String, String> predecessor = new HashMap<>();

        for (String vertex : adjacencyList.keySet()) {
            distance.put(vertex, Integer.MAX_VALUE);
            visited.put(vertex, false);
        }

        distance.put(startVertex, 0);
        priorityQueue.add(new NodeDistance(startVertex, 0));

        while (!priorityQueue.isEmpty()) {
            NodeDistance nodeDistance = priorityQueue.poll();
            String currentVertex = nodeDistance.vertex;

            if (visited.containsKey(currentVertex) && visited.get(currentVertex)) {
                continue;
            }

            visited.put(currentVertex, true);

            if (adjacencyList.containsKey(currentVertex)) {
                for (Edge edge : adjacencyList.get(currentVertex)) {
                    int newDistance = nodeDistance.distance + edge.getWeight();

                    if (newDistance < distance.get(edge.getDestinationId())) {
                        distance.put(edge.getDestinationId(), newDistance);
                        predecessor.put(edge.getDestinationId(), currentVertex);
                        priorityQueue.add(new NodeDistance(edge.getDestinationId(), newDistance));
                    }
                }
            }
        }

        return getShortestPath(startVertex, endVertex, distance, predecessor);
    }

    public Path getShortestPath(String startVertex, String endVertex, Map<String, Integer> distance, Map<String, String> predecessor) {
        LinkedList<String> path = new LinkedList<>();
        String currentVertex = endVertex;

        while (!currentVertex.equals(startVertex)) {
            path.addFirst(currentVertex);
            if (!predecessor.containsKey(currentVertex)) {
                break;
            }
            currentVertex = predecessor.get(currentVertex);
        }

        path.addFirst(startVertex);

        if (distance.containsKey(endVertex) && distance.get(endVertex) != Integer.MAX_VALUE) {
            int shortestDistance = distance.get(endVertex);
            return new Path(path, shortestDistance);
        } else {
            return null;
        }
    }

}

File: Backend_old/src/classes/graph.py
'''classe responsavel por abstrair os nodos e oferecer metodos de arvore '''

import heapq
from functools import reduce

from src.classes.trajeto import Trajeto
from src.classes.edge import Edge


class Graph:
    def __init__(self):
        self.adjacency_list = {}


    def add_edge(self, edge: Edge):
        if edge.source_id not in self.adjacency_list:
            self.adjacency_list[edge.source_id] = []

        if edge.destination_id not in self.adjacency_list:
            self.adjacency_list[edge.destination_id] = []

        self.adjacency_list[edge.source_id].append(edge)

        #Will also add the same path in reverse order, in order for the path to be bidirectional
        self.adjacency_list[edge.destination_id].append(Edge(edge.destination_id, edge.source_id, edge.weight))


    def get_nodes(self):
        return [str(k) for k in self.adjacency_list.keys()]


    def get_graph(self):
        graph_content = {}

        for k in self.adjacency_list.keys():
            if k not in graph_content:
                graph_content[k] = []

            for edge in self.adjacency_list[k]:
                graph_content[k].append(edge.to_dict())

        return graph_content


    def print_graph(self):
        i = 1
        text = ""
        for k in self.adjacency_list.keys():
            for edge in self.adjacency_list[k]:
                text+=f"vertex-{i} {edge.to_string()}\n"
            i+=1

        print(text)

        return text


    def dijkstra(self, start_vertex, end_vertex) -> Trajeto:
        priority_queue = [(0, start_vertex)]
        visited = {}
        distance = {}
        predecessor = {}  # To store the predecessor of each vertex in the shortest path

        for vertex in self.adjacency_list:
            distance[vertex] = float('inf')

        distance[start_vertex] = 0

        while priority_queue:
            current_distance, current_vertex = heapq.heappop(priority_queue)

            if current_vertex in visited:
                continue

            visited[current_vertex] = True

            if current_vertex in self.adjacency_list:
                for edge in self.adjacency_list[current_vertex]:
                    new_distance = current_distance + edge.weight

                    if new_distance < distance[edge.destination_id]:
                        distance[edge.destination_id] = new_distance
                        predecessor[edge.destination_id] = current_vertex  # Update predecessor
                        heapq.heappush(priority_queue, (new_distance, edge.destination_id))

        #Prints the shortest distance from the start vertex to each vertex
        for k in self.adjacency_list.keys():
            print(f"Shortest distance from vertex {start_vertex} to vertex {k}: {distance[k]}")


        print("#"*55)

        trajeto = self.get_shortest_path(start_vertex, end_vertex, distance, predecessor)

        if trajeto:
            self.print_along_with_dashes(
                55,
                f"Shortest path from vertex {start_vertex} to vertex {end_vertex}: {' -> '.join(trajeto.paths)}")
            self.print_along_with_dashes(
                55,
                f"Shortest distance from vertex {start_vertex} to vertex {end_vertex}: {trajeto.distance}")
        else:
            self.print_along_with_dashes(
                55,
                f"Couldn't find a path from {start_vertex} to {end_vertex}!")

        return trajeto


    def print_along_with_dashes(self, number_of_dashes: int, content: str):
        print("#"*number_of_dashes)
        print(content)
        print("#"*number_of_dashes)


    def get_shortest_path(self, start_vertex, end_vertex, distance, predecessor):
        path = []
        current_vertex = end_vertex

        while current_vertex != start_vertex:
            path.insert(0, current_vertex)
            if current_vertex not in predecessor:
                break
            current_vertex = predecessor[current_vertex]

        path.insert(0, start_vertex)

        if end_vertex in distance and distance[end_vertex] != float('inf'):
            shortest_distance = distance[end_vertex]
            return Trajeto(path, shortest_distance)
        else:
            return None